python/cuda_parallel/tests/test_iterators.py

```diff
@@ -0,0 +1,96 @@
+from cuda.parallel.experimental.iterators import (
+    CacheModifiedInputIterator,
+    ConstantIterator,
+    CountingIterator,
+    TransformIterator,
+)
+import cupy as cp
+import numpy as np
+
+
+def test_constant_iterator_equality():
+    it1 = ConstantIterator(np.int32(0))
+    it2 = ConstantIterator(np.int32(0))
+    it3 = ConstantIterator(np.int32(1))
+    it4 = ConstantIterator(np.int64(9))
+
+    assert it1 == it2
+    assert it1 != it3
+    assert it1 != it4
+
+    assert it1.kind == it2.kind == it3.kind
+    assert it1.kind != it4.kind
+
+
+def test_counting_iterator_equality():
+    it1 = CountingIterator(np.int32(0))
+    it2 = CountingIterator(np.int32(0))
+    it3 = CountingIterator(np.int32(1))
+    it4 = CountingIterator(np.int64(9))
+
+    assert it1 == it2
+    assert it1 != it3
+    assert it1 != it4
+
+    assert it1.kind == it2.kind == it3.kind
+    assert it1.kind != it4.kind
+
+
+def test_cache_modified_input_iterator_equality():
+    ary1 = cp.asarray([0, 1, 2], dtype="int32")
+    ary2 = cp.asarray([3, 4, 5], dtype="int32")
+    ary3 = cp.asarray([0, 1, 2], dtype="int64")
+
+    it1 = CacheModifiedInputIterator(ary1, "stream")
+    it2 = CacheModifiedInputIterator(ary1, "stream")
+    it3 = CacheModifiedInputIterator(ary2, "stream")
+    it4 = CacheModifiedInputIterator(ary3, "stream")
+
+    assert it1 == it2
+    assert it1 != it3
+
+    assert it1.kind == it2.kind == it3.kind
+    assert it1.kind != it4.kind
+
+
+def test_equality_transform_iterator():
+    def op1(x):
+        return x
+
+    def op2(x):
+        return 2 * x
+
+    def op3(x):
+        return x
+
+    it = CountingIterator(np.int32(0))
+    it1 = TransformIterator(it, op1)
+    it2 = TransformIterator(it, op1)
+    it3 = TransformIterator(it, op2)
+    it4 = TransformIterator(it, op3)
+
+    assert it1 == it2
+    assert it1 != it3
+    assert it1 == it4
+
+    assert it1.kind == it2.kind == it4.kind
+
+    ary1 = cp.asarray([0, 1, 2])
+    ary2 = cp.asarray([3, 4, 5])
```

Comment:
ary2 = cp.asarray([0, 1, 2])

* We want to exercise corner cases: here we want to be sure we don't compare array elements (arrays may be very large).
* We don't want to give a wrong impression to human readers.

Reply:
Could you please clarify what change you would like to see here? Perhaps an example will help me understand.

Reply:
To clarify, when we previously implemented equality for iterators, `TransformIterator(ary1, op1)` would indeed compare equal to `TransformIterator(ary2, op2)`. But recall that under the hood, `ary1` is converted into a `RawPointer` object and that is what is held by the `TransformIterator`. No elementwise comparison of arrays was done to check for equality.

Reply:
TL;DR: Just keep what you have.

Sorry I forgot to quote the suggested code before:
```
ary2 = cp.asarray([0, 1, 2])
```
This was so that it's immediately obvious that the array elements are not used when determining inequality here:
```
assert it5 != it9
```

But that code is gone now, and you only have `== it8.kind` left. With that it's more obvious that we equality even if the array elements differ.

Reply:
Ah, yes I understand now - thanks!